### DISCUSSION TITLE
Editing hitsounds of nested hit objects

### DISCUSSION THREAD

**Discussion by OliBomby** [ppy/osu#22721] [Ideas]
**Problem**
Currently it is impossible to see or edit the hitsounds of nested hit objects. This is one of the concerns in #12020 and one of the major things in the way of being able to make a complete beatmap in lazer.

I asked peppy about his ideas on this and he had some interesting ideas but I thought it had some issues. He also told me I shouldn't  copy osu! stable, so I thought of some requirements:

- Allow editing of hitsounds of all hitsoundable nested hit objects.
- Show a concise and complete overview of the hitsounds of all hitsoundable nested hit objects.
- It should take no more than 1 second to edit the hitsound of one nested hit object and move on.
- The timeline should not be cluttered too much with hitsound information while not hitsounding.
- Be clear and easy to understand for newbies.

The problem with osu! stable is that it doesn't provide a clear overview of hitsounds in one place because you have to click on each object and open the timing window before you know what sound the object makes. Also the way things are hidden and separated between timing points and hit objects makes it hard for newbies to understand.

**Idea**
My idea is to add a switch to the timeline which switches it to a hitsounding view. In this view you can see all the individual nested hit objects and select them to edit hitsounds. The hit objects which contain the nested objects will be darkened so the nested objects stand out more. This view will also add slider ticks to the timeline so those can be hitsounded. 
On the top there are vertical bars which display the volume for each object. Using the mouse here you can quickly adjust volumes.

Example look: (sorry for the messy editing)
![osu!_A7ADXvuxxa](https://user-images.githubusercontent.com/17460441/220784067-68839110-0e31-49c1-a624-d55921be37fd.png)

I will now argue the requirements this satisfies:
- All hitsoundable nested hit objects can be hitsounded with this solution.
- The timeline provides an overview of the volumes, sample sets, timing, and relation to the hitobjects. It might not be complete because the whistle, finish, clap assignments are not immediately visible, this could be achieved with a piano roll like view but that's a bit overkill I think.
- A doubleclick on an object in the timeline should move to that object and select it, then you can use the hotkeys for whistle, finish, clap to edit the hitsound. This should be very quick to perform. Also editing volumes is just a single click.
- All hitsound information can be removed from the regular timeline so there is no clutter while you are just mapping. I assume mapping and hitsounding are quite separate, so you don't need hitsound information while mapping and vice versa.
- A newbie would only have to find the toggle to switch to hitsounding view in order to understand what is going on. All the hitsounding information is right there, so that should make it easy to experiment with.

**Limitations**
With this idea there will still be no way to change the sampleset of a slider slide sample in the middle of the slider. osu! stable used to be able to controll this with timing points, but lazer doesn't have those for hitsounds anymore. This doesn't get used that often though.

**Comment by Speykious** [comment]
This hitsound view concept seems interesting, but I'm wondering two things:
- What would it look like in other rulesets? Especially mania where hitsounding is made difficult by notes overlapping in the timeline.
- How would hitsound copying work inside the same map?

**Reply by OliBomby** [reply]
> * What would it look like in other rulesets? Especially mania where hitsounding is made difficult by notes overlapping in the timeline.

In mania, the playfield kinda functions as the timeline so the upper timeline is kinda useless anyways. I think hitsounding on the notes in the playfield (similar to how it works in osu! stable) is good enough.

> * How would hitsound copying work inside the same map?

The hitsound information is still stored in the hit objects, so you can copy hitsounds by copying the hit objects.

**Reply by Speykious** [reply]
Working with the playfield for hitsounding in mania is kinda close to what I do for hitsounding, yeah.

For hitsound copying though, the problem is, in a lot of occasions, hitsounds can be the same for different sections where the hit objects are placed differently. Especially in std where they could be timed the same but in a different position. And sure, if we copy the hitsounded hit objects _and then_ change the hit objects' positions, this specific case is not an issue, but it becomes one as soon as the hit objects have different timings or shapes, or as a mapper wants to hitsound their map only after finishing the hit object placement; or even further, if the mapper and the hitsounder are different people (which has been happening recently in my case).

If I want to work solely on hitsounding in Lazer, I don't want to touch hit objects at all, only their hitsound information.

Seeing this hitsound view idea, it makes me think that it might be possible to have a way of selecting hitsounds, then doing usual CTRL+X/C/V operations on them to edit, all without it affecting the placement of hit objects in any way.
(Though then the question is, what happens if the clipboard content doesn't exactly match the hit objects? There could be hitsounds landing on no hit object, in which case it could either disappear or become a storyboard hitsound... Not sure which one is more intuitive.)

Do you think this way of copying hitsounds makes sense for this kind of view?

**Reply by OliBomby** [reply]
Interesting idea for copying hitsounds without their hit objects. I hadn't thought about that before.

You can make some fancy matching algorithm to assign the hitsounds to hit objects with the same timing but it does have that issue of 'what if the hitsound timing doesnt align with the hit objects?' and the UX might be a bit confusing as well. 

Ideally you want to copy hitsounds as if they were entirely separate from the hit objects, so copying hitsounds copies only these 'hitsound objects' which exist on another timeline. If you then copy hitsounds and it doesnt align in timing you can see the issue and manually resnap the hitsound objects. This is similar to having a dedicated hitsound diff for making your hitsounds which is a common practice.

My hitsound view idea is more meant for modding and editing basic hitsounds on the hit object level, so completely separating hitsounds for that is a bit out of scope. I might consider making a dedicated hitsounds editor in the future.

**Reply by Speykious** [reply]
> Ideally you want to copy hitsounds as if they were entirely separate from the hit objects, so copying hitsounds copies only these 'hitsound objects' which exist on another timeline. If you then copy hitsounds and it doesnt align in timing you can see the issue and manually resnap the hitsound objects.

Definitely yeah, though that sounds like it either needs an extension to the existing `.osu` format, or a completely different file to hold that data... I would like that personally but I don't know if it's a good direction overall.

> My hitsound view idea is more meant for modding and editing basic hitsounds on the hit object level, so completely separating hitsounds for that is a bit out of scope.

Hmmm, I'm not sure if creating a separate hitsounding view is that well-adapted for this use-case then... To me it feels like what we have now is enough, but it just needs to display hitsound information for nested hitobjects (especially sliders). Not to mention that there seems to already be elements that would be easy to click and edit.

![image](https://user-images.githubusercontent.com/13885008/220963763-a8839cab-fe77-47bc-8a48-eaead252a2f2.png)
